tools/ci-operator-prowgen/main.go

```diff
@@ -0,0 +1,318 @@
+//   decorate: true
+//   name: branch-ci-<org>-<repo>-<test.Name>
+//   skip_cloning: true
+//   spec: <generatePodSpec>
+func generatePostsubmitForTest(test testDescription, org, repo, branch string, additionalArgs ...string) *prowconfig.Postsubmit {
+	postsubmit := prowconfig.Postsubmit{
+		Agent: "kubernetes",
+		Name:  fmt.Sprintf("branch-ci-%s-%s-%s", org, repo, test.Name),
+		Spec:  generatePodSpec(org, repo, branch, test.Target, additionalArgs...),
+		UtilityConfig: prowconfig.UtilityConfig{
+			DecorationConfig: &prowkube.DecorationConfig{SkipCloning: true},
+			Decorate:         true,
+		},
+	}
+	return &postsubmit
+}
+
+// Given a ci-operator configuration file and basic information about what
+// should be tested, generate a following JobConfig:
+//
+// - one presubmit and one postsubmit for each test defined in config file
+// - if the config file has non-empty `images` section, generate an additinal
+//   presubmit and postsubmit that has `--target=[images]`. This postsubmit
+//   will additionally pass `--promote` to ci-operator and it will have
+//   `artifacts: images` label
+func generateJobs(
+	configSpec *cioperatorapi.ReleaseBuildConfiguration,
+	org, repo, branch string,
+) *prowconfig.JobConfig {
+
+	orgrepo := fmt.Sprintf("%s/%s", org, repo)
+	presubmits := map[string][]prowconfig.Presubmit{}
+	postsubmits := map[string][]prowconfig.Postsubmit{}
+
+	imagesTest := false
+
+	for _, element := range configSpec.Tests {
+		// Check if config file has "images" test defined to avoid name clash
+		// (we generate the additional `--target=[images]` jobs name with `images`
+		// as an identifier, but a user can have `images` test defined in his
+		// config file which would result in a clash)
+		if element.As == "images" {
+			imagesTest = true
+		}
+		test := testDescription{Name: element.As, Target: element.As}
+		presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(test, org, repo, branch))
+		postsubmits[orgrepo] = append(postsubmits[orgrepo], *generatePostsubmitForTest(test, org, repo, branch))
+	}
+
+	if len(configSpec.Images) > 0 && !imagesTest {
+		// TODO: somehow handle the images case better than just not creating this job when there is name conflict
+		test := testDescription{Name: "images", Target: "[images]"}
+		presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(test, org, repo, branch))
+		imagesPostsubmit := generatePostsubmitForTest(test, org, repo, branch, "--promote")
+		imagesPostsubmit.Labels = map[string]string{"artifacts": "images"}
+		postsubmits[orgrepo] = append(postsubmits[orgrepo], *imagesPostsubmit)
+	}
+
+	return &prowconfig.JobConfig{
+		Presubmits:  presubmits,
+		Postsubmits: postsubmits,
+	}
+}
+
+func readCiOperatorConfig(configFilePath string) *cioperatorapi.ReleaseBuildConfiguration {
+	data, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to read ci-operator config (%v)\n", err)
+		os.Exit(1)
```

Comment:
idiomatic nit: Commonly errors are bubbled up to the top-level `main()` which is the only part to use `os.Exit()`. More useful if we're breaking this into multiple files, but might be worth doing now anyway

Reply:
OK